core/apple/test/SourceNSInputStreamTest.kt

```diff
@@ -0,0 +1,354 @@
+                                    input.scheduleInRunLoop(runLoop2, NSDefaultRunLoopMode)
+                                } else if (read >= data.length - 3) {
+                                    // unsubscribe
+                                    input.removeFromRunLoop(runLoop2, NSDefaultRunLoopMode)
+                                }
+                                sink.usePinned {
+                                    val readBytes = input.read(it.addressOf(read).reinterpret(), 3U)
+                                    assertNotEquals(0, readBytes)
+                                    read += readBytes.toInt()
+                                }
+                                if (read == data.length) {
+                                    assertEquals(data, sink.decodeToString())
+                                    completed.unlock()
+                                }
+                            }
+                            NSStreamEventEndEncountered -> fail("$data shouldn't be subscribed")
+                        }
+                    }
+                }
+            }
+            input.scheduleInRunLoop(runLoop1, NSDefaultRunLoopMode)
+            input.open()
+            runBlocking {
+                opened.lockWithTimeout()
+                completed.lockWithTimeout()
+                // wait a bit to be sure delegate is no longer called
+                delay(200)
+            }
+            input.close()
+        }
+
+        consumeSwitching(NSInputStream(data = "default".encodeToByteArray().toNSData()), "default")
+        consumeSwitching(Buffer().apply { writeString("custom") }.asNSInputStream(), "custom")
+        CFRunLoopStop(runLoop1.getCFRunLoop())
+        CFRunLoopStop(runLoop2.getCFRunLoop())
+    }
+
+    @Test
+    fun testSubscribeAfterOpen() {
+        val runLoop = startRunLoop()
+
+        fun subscribeAfterOpen(input: NSInputStream) {
+            println("subscribeAfterOpen(${input::class})")
+            val available = Mutex(true)
+
+            input.delegate = object : NSObject(), NSStreamDelegateProtocol {
+                override fun stream(aStream: NSStream, handleEvent: NSStreamEvent) {
+                    println("${input::class} event = ${handleEvent.asString()} (${NSThread.currentThread.name})")
+                    assertEquals("run-loop", NSThread.currentThread.name)
+                    when (handleEvent) {
+                        NSStreamEventOpenCompleted -> fail("opened before subscribe")
+                        NSStreamEventHasBytesAvailable -> available.unlock()
+                    }
+                }
+            }
+            input.open()
+            input.scheduleInRunLoop(runLoop, NSDefaultRunLoopMode)
+            runBlocking {
+                available.lockWithTimeout()
+            }
+            input.close()
+        }
+
+        subscribeAfterOpen(NSInputStream(data = "default".encodeToByteArray().toNSData()))
```

Comment:
It's reasonable to compare the behavior of custom and default stream implementation during development and debugging, but I don't think we need to run tests for default Foundation streams on a regular basis.